### DISCUSSION TITLE
Tell us your success stories using memray!

### DISCUSSION THREAD

**Discussion by pablogsal** [bloomberg/memray#226] [General]
If you have used `memray` and found it helpful, **please leave a comment below with your success story**. We would love to hear from you and know that our efforts are making a difference. Maybe you used `memray` to detect a memory leak in a C extension or to detect how memory was being used in your large application. If so, you can tell us how did `memray` help! 

Knowing how we managed to help your use case will help us track trends and target areas for improvement, prioritize new features and development and identify potential bugs or areas of confusion.

Thank you for taking the time to provide feedback! ❤️

**Comment by JelleZijlstra** [comment]
I maintain a large Python application at Quora. It runs a couple of pre-warming tasks at startup to improve performance at runtime. I used memray to figure out which parts of the warmup sequence were using memory. Fixing a few of the bottlenecks helped us achieve a significant cost reduction.

**Comment by tonybaloney** [comment]
I work on a [Python C extension](https://github.com/microsoft/picologging) that is designed for performance. I successfully used memray to identify memory leaks (refcount leaks) of Python objects created in the extension module whilst developing it. We're using memray for Quality Assurance testing on this library that's being used in a hot-loop executed billions of times a day. 
All the issues identified were done during the development cycle instead of identifying memory usage in production environments.

**Reply by pamelafox** [reply]
Pics for proof! (From that project)

<img width="1720" alt="Memray diagram before" src="https://user-images.githubusercontent.com/297042/195714426-f4d2712d-05c3-43cc-b852-18d0dabce28a.png">

<img width="1722" alt="Memray diagram after" src="https://user-images.githubusercontent.com/297042/195714432-87e224d7-18b7-4b2b-9280-42917d1ea47d.png">

**Comment by jmg-duarte** [comment]
I work on a data pipeline system (akin to Prefect, Dagster, Airflow, etc) and out of the blue, the system worker started crashing due to OOM errors. While we were able to identify that it was a memory leak on the LZ4 part of the system, we couldn't track down whether it was on the python-lz4 C bindings or the actual LZ4 library. 
With the collaboration of the LZ4's author, I isolated the cause to python-lz4 C bindings (https://github.com/lz4/lz4/issues/1096). 
Afterwards, using memray (with the native flag) I was able to *directly* track the line that allocated the leak memory - memory wasn't being freed in the error path - adding a `PyMem_Free`  resolved the issue (https://github.com/python-lz4/python-lz4/pull/255).

**Comment by bgyori** [comment]
I lead the development of several open-source scientific software projects for computational systems biology. We recently used memray to profile the memory usage of our [Gilda](https://github.com/indralab/gilda) software and successfully identified where it can be streamlined. I found the live monitoring as well as the HTML-based visualization features of memray particularly useful.

**Comment by hmcezar** [comment]
I work on the development of [HyMD](https://github.com/Cascella-Group-UiO/HyMD), a scientific software for molecular dynamics simulations using a hybrid particle-field approach. Using `memray` we were able to track a huge memory leak from OpenMPI that leaked into our application, causing many simulations to crash after a while. `memray` was an indispensable tool to help us fix this problem.

**Comment by pquentin** [comment]
I co-maintain urllib3, the package with the most downloads on PyPI. Anything we do has a big impact, for better or for worse. As part of urllib3 2.0, we wanted to fix how read() works with compressed data to be [more intuitive](https://github.com/urllib3/urllib3/issues/709). This required adding a buffer, but our first attempt [doubled the memory usage](https://github.com/urllib3/urllib3/pull/2712#issuecomment-1308691054). Armed with memray and its intuitive flamegraphs, I managed to come up with a [solution](https://github.com/urllib3/urllib3/issues/2787#issuecomment-1311254728) that used the same memory as before. Without that reduction we could not have shipped this improvement. And with pytest-memray we're now confident that there won't be any regressions!

**Comment by farhanhubble** [comment]
I was handed over a massive natural language processing pipeline that was consuming a whopping 250GB of RAM!! I looked at the models. Quite a few of them but the largest one was about 1GB. I tried `memray` and looked at the summary. The peak memory utilization was just a few GBs. This led me to believe that something was allocating and deallocating memory too fast and the Python interpreter was not able to return the memory back to the system in time. Eventually found and fixed the code that was repeatedly loading a machine learning model on every call. 

Memray is fast, beautiful and accurate. I wish I could do a t-shirt with a snapshot of `memray --live` to thank the devs.

**Comment by maartenbreddels** [comment]
I've described at https://github.com/widgetti/solara/discussions/172 how I used memray to find the area where a memory leak was in Solara and/or Reacton, found the memory leak (global reference) using [objgraph](https://objgraph.readthedocs.io/en/stable/) and then confirm the fix using memray again.

**Comment by ankush** [comment]
Memray played a huge role in our effort to reduce memory footprint at @frappe. I've used many different memory profilers in past but none come close in terms of usability and simplicity. We were able to slash anywhere from 10-35% memory usage (depending on the configs) and that translates to gigabytes of savings across servers that serve 6k+ sites :smile: 

`--follow-fork` was quite useful feature when dealing with worker processes that use forking.

references:
https://github.com/frappe/frappe/pull/21467
https://github.com/frappe/frappe/pull/21473
https://github.com/frappe/frappe/pull/21474
https://github.com/frappe/frappe/pull/21475

**Comment by tvalentyn** [comment]
Memray was instrumental in pinpointing a source of a tricky memory leak in Apache Beam that I was investigating: https://github.com/apache/beam/issues/28246.  I tried several tools and Memray is by far the fastest and the most usable tool I tried. I haven't used memray before and originally  tried `memray summary` command which wasn't informative, and made me  dismiss Memray. I decided to give Memray another try a couple of weeks later after very frustrating experiences with many other, mostly older tools, and discovered that `memray table` command   does wonders for hunting down my leak. I later used Memray to simplify the repro and Memray dramatically reduced my feedback loop during debugging.

**Comment by terapyon** [comment]
I wrote an article about Memray in Japanese last year. Title is "メモリプロファイラ「Memray」の解説", it is "Explanation of the memory profiler “Memray". https://gihyo.jp/article/2023/06/monthly-python-2306
This is one of famous online media in Japan. 

Then, I gave a talk it at PyCon APAC 2023 in Japanese.
https://2023-apac.pycon.jp/timetable?id=FHRP3H

**Comment by tojimahammatov** [comment]
I work on developing services which some of them perform AI model inference. We noticed a memory leak issue in one service and detected 2 reasons for it. Although we fixed one, tracing the other issue was quite difficult. Memray helped us to identify the bug accurately by providing nice reports.

**Comment by alonmr** [comment]
I am working on a python service and we noticed a memory leak on scale systems. I integrated memray to deliver fine grained memory profiling and it helped us understand 2 things:
1. The resident memory is leaking and the heap memory is not.
2. We are using memory inefficiently at some places.

With memray reports we came to a conclusion we had severe memory fragmentation so we optimized our memory usage and it worked wonders.
I even gave a presentation for the R&D team on how we did it.
Thanks for this amazing tool!

**Comment by rbetts** [comment]
After chasing an out-of-memory issue all afternoon, I found memray and it immediately showed me I was looking in completely the wrong place. Worked out of the box .. simple and powerful. Thanks so much for making this tool! I especially appreciate that memray profiles without requiring (manually) changing code to add instrumentation or profile points.

**Comment by jvns** [comment]
One of the management scripts in my Django web app was using more memory than I expected, and I used memray to figure out that the management script was importing a big dependency that it didn't need to! It was very easy to use and I really appreciated that `memray run` prints out the appropriate `memray flamegraph` command to run.
